File: frontend/src/services/trading.ts
import { apiClient } from './api'
import type { Order, Trade, Position, PortfolioData, ExchangeAccount, ApiResponse, PaginatedResponse } from '../types'

export const tradingService = {
  async createOrder(data: {
    symbol: string
    side: 'buy' | 'sell'
    type: 'market' | 'limit' | 'stop'
    quantity: number
    price?: number
    is_simulation?: boolean
    signal_id?: number
    exchange_account_id?: number
  }): Promise<ApiResponse<Order>> {
    const response = await apiClient.post('/trading/orders', data)
    return response.data
  },

  async cancelOrder(orderId: number): Promise<ApiResponse<Order>> {
    const response = await apiClient.post(`/trading/orders/${orderId}/cancel`)
    return response.data
  },

  async getOrders(params?: {
    status?: string
    symbol?: string
    is_simulation?: boolean
    page?: number
    per_page?: number
  }): Promise<ApiResponse<PaginatedResponse<Order>>> {
    const response = await apiClient.get('/trading/orders', { params })
    return response.data
  },

  async getOrderDetail(orderId: number): Promise<ApiResponse<Order>> {
    const response = await apiClient.get(`/trading/orders/${orderId}`)
    return response.data
  },

  async getPositions(params?: {
    symbol?: string
    is_simulation?: boolean
  }): Promise<ApiResponse<Position[]>> {
    const response = await apiClient.get('/trading/positions', { params })
    return response.data
  },

  async getTrades(params?: {
    symbol?: string
    is_simulation?: boolean
    start_date?: string
    end_date?: string
    page?: number
    per_page?: number
  }): Promise<ApiResponse<PaginatedResponse<Trade>>> {
    const response = await apiClient.get('/trading/trades', { params })
    return response.data
  },

  async getPortfolio(is_simulation: boolean = true): Promise<ApiResponse<PortfolioData>> {
    const response = await apiClient.get('/trading/portfolio', { params: { is_simulation } })
    return response.data
  },

  async getExchangeAccounts(): Promise<ApiResponse<ExchangeAccount[]>> {
    const response = await apiClient.get('/trading/exchange-accounts')
    return response.data
  },

  async createExchangeAccount(data: {
    exchange: string
    api_key: string
    api_secret: string
    passphrase?: string
    is_sandbox?: boolean
    name?: string
  }): Promise<ApiResponse<ExchangeAccount>> {
    const response = await apiClient.post('/trading/exchange-accounts', data)
    return response.data
  },

  async deleteExchangeAccount(accountId: number): Promise<ApiResponse<{ message: string }>> {
    const response = await apiClient.delete(`/trading/exchange-accounts/${accountId}`)
    return response.data
  },

  async getMarketPrice(symbol: string): Promise<ApiResponse<{ symbol: string; price: number }>> {
    const response = await apiClient.get(`/trading/market/price/${symbol}`)
    return response.data
  },

  async getAllMarketPrices(): Promise<ApiResponse<Record<string, number>>> {
    const response = await apiClient.get('/trading/market/prices')
    return response.data
  }
}

File: server/routes/trading.py
from flask import Blueprint, request, jsonify
from middleware.auth import require_auth
from services import trading_service

trading_bp = Blueprint('trading', __name__, url_prefix='/api/trading')


@trading_bp.route('/orders', methods=['POST'])
@require_auth
def create_order():
    user_id = request.current_user_id
    data = request.get_json()
    result = trading_service.create_order(user_id, data)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/orders/<int:order_id>/cancel', methods=['POST'])
@require_auth
def cancel_order(order_id):
    user_id = request.current_user_id
    result = trading_service.cancel_order(user_id, order_id)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/orders', methods=['GET'])
@require_auth
def get_orders():
    user_id = request.current_user_id
    params = request.args.to_dict()
    result = trading_service.get_orders(user_id, params)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/orders/<int:order_id>', methods=['GET'])
@require_auth
def get_order_detail(order_id):
    user_id = request.current_user_id
    result = trading_service.get_order_detail(user_id, order_id)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/positions', methods=['GET'])
@require_auth
def get_positions():
    user_id = request.current_user_id
    params = request.args.to_dict()
    result = trading_service.get_positions(user_id, params)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/trades', methods=['GET'])
@require_auth
def get_trades():
    user_id = request.current_user_id
    params = request.args.to_dict()
    result = trading_service.get_trades(user_id, params)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/portfolio', methods=['GET'])
@require_auth
def get_portfolio():
    user_id = request.current_user_id
    is_simulation = request.args.get('is_simulation', 'true').lower() == 'true'
    result = trading_service.get_portfolio(user_id, is_simulation)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/exchange-accounts', methods=['GET'])
@require_auth
def get_exchange_accounts():
    user_id = request.current_user_id
    result = trading_service.get_exchange_accounts(user_id)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/exchange-accounts', methods=['POST'])
@require_auth
def create_exchange_account():
    user_id = request.current_user_id
    data = request.get_json()
    result = trading_service.create_exchange_account(user_id, data)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/exchange-accounts/<int:account_id>', methods=['DELETE'])
@require_auth
def delete_exchange_account(account_id):
    user_id = request.current_user_id
    result = trading_service.delete_exchange_account(user_id, account_id)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/market/price/<symbol>', methods=['GET'])
def get_market_price(symbol):
    result = trading_service.get_market_price(symbol)
    return jsonify(result), result.get('code', 200)


@trading_bp.route('/market/prices', methods=['GET'])
def get_all_market_prices():
    result = trading_service.get_all_market_prices()
    return jsonify(result), result.get('code', 200)
